From e7865ef80de5c7d690567fba4d59ff596b1e341c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?M=C3=A5rten=20Kongstad?= <amhk@google.com>
Date: Thu, 19 Feb 2026 20:05:00 +0000
Subject: [PATCH 1/3] derive_sdk: introduce C extension

Note: no new modules opted in to SDK extensions as part of C.

Bug: 480974361
Test: atest derive_sdk_test
Flag: NONE platform SDK finalization
Change-Id: Ia14de0a8a5347373003f6b2516657ee7da294c6d
---
 derive_sdk/derive_sdk.cpp      | 11 ++++++++++-
 derive_sdk/derive_sdk_test.cpp |  9 +++++++++
 2 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/derive_sdk/derive_sdk.cpp b/derive_sdk/derive_sdk.cpp
index 65b061ffb52..a38abc32f67 100644
--- a/derive_sdk/derive_sdk.cpp
+++ b/derive_sdk/derive_sdk.cpp
@@ -75,12 +75,14 @@ static const std::unordered_set<SdkModule> kVModules = {};
 
 static const std::unordered_set<SdkModule> kBModules = {SdkModule::NEURAL_NETWORKS};
 
+static const std::unordered_set<SdkModule> kCModules = {};
+
 static const std::string kSystemPropertiesPrefix = "build.version.extensions.";
 
 void ReadSystemProperties(std::map<std::string, std::string>& properties) {
   const std::string default_ = "<not set>";
 
-  for (const auto& dessert : {"r", "s", "t", "ad_services", "u", "v", "b"}) {
+  for (const auto& dessert : {"r", "s", "t", "ad_services", "u", "v", "b", "c"}) {
     properties[kSystemPropertiesPrefix + dessert] =
         android::base::GetProperty(kSystemPropertiesPrefix + dessert, default_);
   }
@@ -239,6 +241,13 @@ bool SetSdkLevels(const std::string& mountpath) {
     }
   }
 
+  relevant_modules.insert(kCModules.begin(), kCModules.end());
+  if (android::modules::sdklevel::IsAtLeastC()) {
+    if (!GetAndSetExtension("c", db, relevant_modules, versions)) {
+      return false;
+    }
+  }
+
   // Consistency check: verify all modules with requirements is included in some dessert
   for (const auto& ext_version : db.versions()) {
     for (const auto& requirement : ext_version.requirements()) {
diff --git a/derive_sdk/derive_sdk_test.cpp b/derive_sdk/derive_sdk_test.cpp
index 54b8cdeadf4..47dfae518b5 100644
--- a/derive_sdk/derive_sdk_test.cpp
+++ b/derive_sdk/derive_sdk_test.cpp
@@ -59,6 +59,9 @@
 // Only expect the B extension level to be set on B+ devices.
 #define EXPECT_B(n) EXPECT_EQ(GetB(), android::modules::sdklevel::IsAtLeastB() ? (n) : -1)
 
+// Only expect the C extension level to be set on C+ devices.
+#define EXPECT_C(n) EXPECT_EQ(GetC(), android::modules::sdklevel::IsAtLeastC() ? (n) : -1)
+
 class DeriveSdkTest : public ::testing::Test {
  protected:
   void TearDown() override { android::derivesdk::SetSdkLevels("/apex"); }
@@ -115,6 +118,8 @@ class DeriveSdkTest : public ::testing::Test {
 
   int GetB() { return android::base::GetIntProperty("build.version.extensions.b", -1); }
 
+  int GetC() { return android::base::GetIntProperty("build.version.extensions.c", -1); }
+
   void EXPECT_ADSERVICES(int n) {
     int actual = android::base::GetIntProperty("build.version.extensions.ad_services", -1);
     // Only expect the AdServices extension level to be set on T+ devices.
@@ -349,6 +354,10 @@ TEST_F(DeriveSdkTest, Baklava) {
   EXPECT_B(1);
 }
 
+TEST_F(DeriveSdkTest, CinnamonBun) {
+  // Nothing to do: no new modules were added in B
+}
+
 int main(int argc, char** argv) {
   ::testing::InitGoogleTest(&argc, argv);
   return RUN_ALL_TESTS();
-- 
2.53.0.345.g96ddfc5eaa-goog

